Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/TfprPrefetchingStrategy.java

```diff
@@ -0,0 +1,344 @@
+package nl.vu.cs.s2group.nappa.prefetch;
+
+import android.app.Activity;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import nl.vu.cs.s2group.nappa.graph.ActivityNode;
+import nl.vu.cs.s2group.nappa.util.NappaUtil;
+
+/**
+ * This strategy employs a Link Analysis approach implementing a PageRank-based algorithm
+ * using the time a user spends in the activities and how frequent the user access the
+ * activities to decide which nodes to select.
+ * <p>
+ * Only a subgraph of the ENG is considered in the calculations. The score calculations are
+ * performed at runtime and are not persisted in the database.
+ * <p>
+ * Strategy inspired on the paper Personalized PageRank for Web Page Prediction Based
+ * on Access Time-Length and Frequency from 2007.
+ * <p>
+ * This strategy accepts the following configurations:
+ * <ul>
+ * </ul>
+ *
+ * @see <a href="https://dl.acm.org/doi/10.1109/WI.2007.145">Personalized PageRank paper</a>
+ */
+public class TfprPrefetchingStrategy extends AbstractPrefetchingStrategy {
+    private static final String LOG_TAG = TfprPrefetchingStrategy.class.getSimpleName();
+
+    public TfprPrefetchingStrategy() {
+        super();
+
+        /*
+          On the testing phase, the TFPR score was rather low.
+          The TFPR score was higher for pages with longer time (e.g. 30+ seconds) and
+          lower for pages with short time (e.g., 5- seconds).
+          Low TFPR scores were always below 10%, often below 5%.
+          High TFPR scores were always between 40% to 50~55%.
+          While testing, there was only 2 occasions where a page obtained a TFPR score equal
+          or higher than 60%, which is the default lower threshold for the weight score:
+
+          * The long time was increased to a few minutes;
+          * The current node had a single child only;
+
+          In both cases the TFPR score was between 60% and 75%.
+
+          Since this is not always the case, it seems reasonable to reduce the lower threshold.
+          The value is only overridden if it was not already overridden using the user-defined
+          configurations.
+         */
```

Comment:
There is the question if this behaviour will occur during the experiments and/or in real usage scenario,

Reply:
You can try with a single run of the experiment over all subjects and see how the scores fluctuate, ok?